stan/math/opencl/kernel_generator/operation.hpp

```diff
@@ -0,0 +1,226 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_OPERATION_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_OPERATION_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/is_valid_expression.hpp>
+#include <stan/math/opencl/matrix_cl_view.hpp>
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/kernel_cl.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <cl.hpp>
+#include <algorithm>
+#include <string>
+#include <tuple>
+#include <set>
+#include <array>
+#include <numeric>
+
+namespace stan {
+namespace math {
+
+/**
+ * Parts of an OpenCL kernel, generated by an expression
+ */
+struct kernel_parts {
+  std::string body;  // the body of the kernel - code executing operations
+  std::string args;  // kernel arguments
+};
+
+/**
+ * Base for all kernel generator operations.
+ * @tparam Derived derived type
+ * @tparam ReturnScalar scalar type of the result
+ * @tparam Args types of arguments to this operation
+ */
+template <typename Derived, typename ReturnScalar, typename... Args>
+class operation : public operation_base {
+ public:
+  static_assert(conjunction<std::is_base_of<
+                    operation_base, std::remove_reference_t<Args>>...>::value,
+                "operation: all arguments to operation must be operations!");
+  // number of arguments this operation has
+  static constexpr int N = std::tuple_size<std::tuple<Args...>>::value;
+  // value representing a not yet determined size
+  static const int dynamic = -1;
+
+  /**
+   * Constructor
+   * @param arguments Arguments of this expression that are also valid
+   * expressions
+   */
+  explicit operation(Args&&... arguments)
```

Comment:
That is fine, this is a constructor.

Reply:
> That is fine, this is a constructor.

I think it's not fine _because_ it's a constructor. The reference type of the class template parameters during initialization can differ from the reference type used in the constructor. link below shows what I'm talking about. I think you need to do something like the below to check the decayed types match

```cpp
template <typename... Args_, require_same_t<Args, Args_>...> // I think that will work?
operation(Args_&& arguments) {...}
```

https://godbolt.org/z/EQFvFm

Reply:
Your godbolt example does not work, because you made an error explicitely setting template parameters. This works: https://godbolt.org/z/uxJhq3. And so does kernel generator - a similar example is actually in tests - in `binary_operation_test.cpp`: `multiple_operations_accepts_lvalue`.